### PR TITLE
Allow for dynamic metric name

### PR DESCRIPTION
**Pull request by erickpintor**
Hi folks,

I'm opening this PR as an attempt to get this feature merged since I need it in one of my projects. It was originally implemented by @drawks at the PR #89. I've reimplemented the same feature on top of the most recent version of the master branch and added a few test cases.

I hope this helps.

### CODE REVIEW

**Comment by drawks** [comment]
documentation needs a bit of caveat about colliding names; i.e. each metric can only have one unique set of labels.

File: README.md [hunk 1]
```diff
@@ -123,7 +123,17 @@ follows:
     test.web-server.foo.bar
      => test_web_server_foo_bar{}
 
-Each mapping in the configuration file must define a `name` for the metric.
+Each mapping in the configuration file must define a `name` for the metric. The
+metric's name can contain `$n`-style references to be replaced by the n-th
+wildcard match in the matching line. That allows for dynamic rewrites, such as:
```
Comment (matthiasr) [inline README.md]: Can you also mention how this works with the regex match type?

**Comment by erickpintor** [comment]
Sure. Could you please clarify what happens when the metric names collide? I just want to make sure the docs are accurate.

**Comment by matthiasr** [comment]
If two metrics have the same name, then they also must have the same label
set (set of label names, they can have different values of course).
Otherwise the /metrics endpoint will throw errors. That's already the case
but the risk is higher that it might happen accidentally with templating.

**Comment by erickpintor** [comment]
@matthiasr could you take another pass. I'm not sure if the docs are clear enough.

File: README.md [hunk 2]
```diff
@@ -123,7 +123,32 @@ follows:
...
+mappings:
+- match: test\.(\w+)\.(\w+)\.counter
+  match_type: regex
+  name: "${2}_counter"
+  labels:
+    provider: "$1"
+```
+
+Please note that metrics with the same name must also have the same set of
+labels names.
```
Comment (matthiasr) [inline README.md]: there's a mini typo here: "labels names" – I can fix that up myself though.

**Comment by erickpintor** [comment]
By the way, I was using the docker image before finding out I needed this feature. Do you know by any chance when we can get a new version of the docker image?

**Comment by matthiasr** [comment]
It should be available at `prom/statsd-exporter:master` in a few minutes. I would appreciate if you could test it under real-life conditions before I make a release.

**Comment by matthiasr** [comment]
(this doesn't need to be a very extended test, if it works for you I'd make a release tomorrow or so)

**Comment by erickpintor** [comment]
Sure. I can test the master tag.
I'm already using that code in a docker container, I'm just using my fork instead of the official image.

**Comment by matthiasr** [comment]
Alright then, release time!

**Pull request merged**